src/Disciplina/Launcher/Mode.hs

```diff
@@ -1,49 +1,69 @@
-{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TemplateHaskell        #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
 
--- | Module contains the definition of WorkMode and its implementations.
+-- | Module contains the definition of basic WorkMode and its implementations.
 
 module Disciplina.Launcher.Mode
        (
          -- * Constraints
          BasicWorkMode
 
          -- * Implementations
-       , BasicRealMode
-       , BasicRealContext (..)
+       , RealMode
+       , NodeContext (..)
+       , CustomContext
+       , ncCustomCtx
+
+       , -- * Misc
+         FormNodeContext (..)
        ) where
 
 import Universum
 
 import Control.Lens (makeLenses)
-import Mockable (MonadMockable, Production (..))
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Disciplina.Async (Forall, Pure)
 import System.Wlog (HasLoggerName (..), LoggerName, WithLogger)
 
-import Disciplina.DB.Class (MonadDB)
-
 ---------------------------------------------------------------------
 -- WorkMode classes
 ---------------------------------------------------------------------
 
 -- | Set of typeclasses which define basic capabilities of Disciplina node
 type BasicWorkMode m =
     ( WithLogger m
-    , MonadMockable m
     , MonadIO m
-    , MonadDB m
+    , MonadBaseControl IO m
+    , Forall (Pure m)  -- allows to use lifted-async
     )
 
 ---------------------------------------------------------------------
 -- WorkMode implementations
 ---------------------------------------------------------------------
 
-data BasicRealContext = BasicRealContext
-    { _brcLoggerName :: LoggerName
+-- | For node role returns related context.
+type family CustomContext r = cc | cc -> r
+
+-- | Contains basic context suitable for any node + custom context,
+-- which depends on node role.
+data NodeContext r = NodeContext
+    { _ncCustomCtx  :: CustomContext r
+    , _ncLoggerName :: LoggerName
     }
 
-makeLenses ''BasicRealContext
+makeLenses ''NodeContext
+
+type RealMode r = ReaderT (NodeContext r) IO
+
+instance {-# OVERLAPPING #-} HasLoggerName (RealMode r) where
+    askLoggerName = view ncLoggerName
+    modifyLoggerName name = local $ ncLoggerName %~ name
+
+---------------------------------------------------------------------
+-- Misc
+---------------------------------------------------------------------
 
-type BasicRealMode = ReaderT BasicRealContext Production
+-- | Forms context using given pack of resources.
+class FormNodeContext res ctx | ctx -> res, res -> ctx where
+    formNodeContext :: res -> IO ctx
```

Comment:
Frankly, this typeclass seemingly doesn't add a lot of convenience - it basically has only two instances (and will almost always have), and for these two instances it would be quite straightforward to just have two functions `formEducatorNodeContext` and `formWitnessNodeContext`, which you'll be able to pass as a param to `runRealMode`.
(I don't mean to be like Vieth, but this class really seems to complicate the code more than simplify it.)

Reply:
Okay, I was just too lazy to create new two modules, will do now

Reply:
@Martoon-00 didn't get it, which two new _modules_? :thinking:

Reply:
Oh not, I can just place those functions near to resources types, this typeclass really doesn't make sense

Reply:
Sorry, just saw your comment. I thought dedicated `runEducatorMode` should be places in separate `Disciplina.Educator.Launcher.Runner` module (and I did so eventually)